app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,174 @@
+package org.commcare.dalvik.geo;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * Allows evaluation contexts to support the here() XPath function,
+ * which returns the current location.
+ *
+ * In addition, an EntitySelectActivity can register itself to be refreshed whenever
+ * a new value of here() is obtained (whenever the location changes).
+ *
+ * No locations are requested if here() is never evaluated.
+ *
+ * @author Forest Tong
+ */
+public class HereFunctionHandler implements IFunctionHandler, LocationListener {
+    public static final String HERE_NAME = "here";
+    private GeoPointData location;
+
+    private LocationManager mLocationManager;
+
+    private boolean requestingLocationUpdates;
+    private boolean locationGoodEnough;
+
+    private final Context context = CommCareApplication._().getApplicationContext();
+
+    // If there are more general uses for HereFunctionHandler, the type of this field can be
+    // generalized to a listener interface.
+    private EntitySelectActivity entitySelectActivity;
+
+    public HereFunctionHandler() {}
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    public void unregisterEntitySelectActivity() {
+        this.entitySelectActivity = null;
+    }
+
+    // The EntitySelectActivity must subscribe before this method is called if a fresh location is desired.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if (entitySelectActivity != null && !entitySelectActivity.containsHereFunction) {
+            refreshLocation();
+            allowGpsUse();
+            entitySelectActivity.containsHereFunction = true;
+        }
+        if (location == null) {
+            return "";
+        }
+        return location.getDisplayText();
+    }
+
+    public void allowGpsUse() {
+        if (!locationGoodEnough && !requestingLocationUpdates) {
+            requestLocationUpdates();
+        }
+    }
+
+    public void forbidGpsUse() {
+        if (requestingLocationUpdates) {
+            removeLocationUpdates();
+        }
+    }
```

Comment:
I considered wording like `pauseGpsUse` or `resumeGpsUse,` but decided that these don't fit, because what I'm trying to do with this interface is have the client only worry about whether GPS/network use is permitted, and not have to worry about whether the HereFunctionHandler is actually using GPS/network. For example, if the HereFunctionHandler finished finding an accurate enough location, it would be misleading to say `resumeGpsUse` when you mean to say "use GPS if you still need it." If you think allow/forbid is unclear, I could consider permit/prohibit, or enable/disable, or something similar.

The reason `allowGpsUse` can't be moved into `registerEntitySelectActivity` is that we want to register an `EntitySelectActivity` even if it doesn't contain the `here` function (and if there's no `here` function, GPS use is forbidden).

I like the suggestion of renaming it to `registerEvalLocationListener.`